parse.go

```diff
@@ -0,0 +1,419 @@
+		}
+	}
+
+	// TODO: error. Reason: no expression found.
+	return LocalDeclaration{}
+}
+
+// ---------------------------------------------------------------------
+
+func (p *parser) parseMatcher() Matcher {
+	var matcher Matcher
+
+	for itm := p.current(); itm.typ != itemEOF; itm = p.next() {
+		// TODO: Error handling: unexpected token
+		//nolint:exhaustive
+		switch itm.typ {
+		case itemExpressionOpen:
+			matcher.MatchStatements = append(matcher.MatchStatements, p.parseExpression())
+		case itemCatchAllKey, itemNumberLiteral, itemQuotedLiteral, itemUnquotedLiteral:
+			matcher.Variants = append(matcher.Variants, Variant{
+				Keys: p.parseVariantKeys(),
+				QuotedPattern: QuotedPattern{
+					Patterns: p.parsePatterns(),
+				},
+			})
+		}
+	}
+
+	return matcher
+}
+
+func (p *parser) parseVariantKeys() []VariantKey {
+	var keys []VariantKey
+
+	for itm := p.current(); itm.typ != itemQuotedPatternOpen; itm = p.next() {
+		// TODO: Error handling: unexpected token
+		//nolint:exhaustive
+		switch itm.typ {
+		case itemCatchAllKey:
+			keys = append(keys, CatchAllKey{})
+		case itemNumberLiteral, itemQuotedLiteral, itemUnquotedLiteral:
+			keys = append(keys, LiteralKey{Literal: p.parseLiteral()})
+		}
+	}
+
+	return keys
+}
+
+func (p *parser) parseOption() Option { //nolint:ireturn
+	var identifier Identifier
+
+	for itm := p.current(); itm.typ != itemExpressionClose; itm = p.next() {
+		// TODO: Error handling: unexpected token
+		//nolint:exhaustive
+		switch itm.typ {
+		case itemOption:
+			identifier = p.parseIdentifier()
+
+		case itemNumberLiteral, itemQuotedLiteral, itemUnquotedLiteral:
+			return LiteralOption{Literal: p.parseLiteral(), Identifier: identifier}
+
+		case itemVariable:
+			return VariableOption{
+				Variable:   Variable(p.current().val[1:]), // omit "$" prefix //TODO: Lexer should not capture variable prefix
+				Identifier: identifier,
+			}
+		}
+	}
+
+	// todo: error. Reason: value is missing for option
+	return nil
+}
+
+func (p *parser) parseLiteral() Literal { //nolint:ireturn
+	// TODO: Error handling: unexpected token
+	//nolint:exhaustive
+	switch itm := p.current(); itm.typ {
+	case itemNumberLiteral:
+		var num float64
+		if err := json.Unmarshal([]byte(itm.val), &num); err != nil {
+			// TODO: Return error instead of panic
+			panic(err)
+		}
+
+		return UnquotedLiteral{Value: NumberLiteral(num)}
+	case itemQuotedLiteral:
+		return QuotedLiteral(p.current().val)
+	case itemUnquotedLiteral:
+		return UnquotedLiteral{Value: NameLiteral(p.current().val)}
+	}
+
+	// TODO: error. Reason: unexpected token
+	return nil
+}
+
+func (p *parser) parseFunction() Function {
+	return Function{Prefix: rune(p.current().val[0]), Identifier: p.parseIdentifier()}
+}
+
+func (p *parser) parseIdentifier() Identifier {
+	full := strings.Split(p.current().val, ":")
+
+	var (
+		ns   string
+		name string
+	)
+
+	//nolint:gomnd
+	// TODO: error handling: case unexpected length
```

Comment:
The lexer will error on the invalid function name, you don't have to worry about it.